### PR TITLE
Add support for dplyr::arrange

### LINKED ISSUE #47: arrange is not supported
Considering `arrange()` is part of dplyr and tidyr I would expected support for it. For example:
```
iris %>%
  group_by(Species) %>%
  summarise(mean_Sepal_Length = mean(Sepal.Length), .groups = "drop") %>%
  arrange(desc(mean_Sepal_Length))
```
now gives:

```
group_by: one grouping variable (Species)
summarise: now 3 rows and 2 columns, ungrouped
```

while I would expect something like

```
group_by: one grouping variable (Species)
summarise: now 3 rows and 2 columns, ungrouped
arrange: now arranged in descending order by one variable (mean_Sepal_Length)
```

I think that would be a nice addition, if it would fit within the logic of this package.

### PR DESCRIPTION
**Pull request by orgadish**
Fixes #47 

Adds support for `dplyr::arrange`, specifically:
1. Displaying variable names passed in.
2. Including modifiers, e.g. `desc(carb)`
3. Unpacking dynamic selections via `across()` and `pick()`
4. Indicating when it is grouped or not.

``` r
tidylog::arrange(mtcars, across(everything())) |> head(3)
#> arrange: sorted rows by mpg, cyl, disp, hp, drat, …
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Camaro Z28          13.3   8  350 245 3.73 3.840 15.41  0  0    3    4

tidylog::arrange(mtcars, across(starts_with("c"))) |> head(3)
#> arrange: sorted rows by cyl, carb
#>                 mpg cyl  disp hp drat    wt  qsec vs am gear carb
#> Datsun 710     22.8   4 108.0 93 3.85 2.320 18.61  1  1    4    1
#> Fiat 128       32.4   4  78.7 66 4.08 2.200 19.47  1  1    4    1
#> Toyota Corolla 33.9   4  71.1 65 4.22 1.835 19.90  1  1    4    1

tidylog::arrange(dplyr::group_by(mtcars, cyl), carb, desc(mpg)) |> head(3)
#> arrange (grouped): sorted rows within groups by carb, desc(mpg)
#> # A tibble: 3 × 11
#> # Groups:   cyl [1]
#>     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1  33.9     4  71.1    65  4.22  1.84  19.9     1     1     4     1
#> 2  32.4     4  78.7    66  4.08  2.2   19.5     1     1     4     1
#> 3  27.3     4  79      66  4.08  1.94  18.9     1     1     4     1
```

<sup>Created on 2026-06-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

See tests for more examples.

### CODE REVIEW

**Comment by orgadish** [comment]
@elbersb I figured I'd use this as a true test case of adding a new function in the new system.

**Comment by elbersb** [comment]
If this is showing the variables it sorts by, don't we know all of them for the NA functionality that was proposed in the issue?

**Comment by orgadish** [comment]
Yeah you're right. I was running into issues with expressions like `desc()` but since I'm already unpacking the other cases, I can do that for desc.

And I can also catch cases like `arrange(x, c(A, B, C))` and `arrange(x, A:C)`

**Comment by orgadish** [comment]
@elbersb to catch the remaining dplyr cases (enumerated extensively in `test_arrange`) and also made sure that it now logs for NAs.

Do you thoughts on what the NA message should say? Right now I have it as 
`"some columns contained NAs which sort last ({format_list(cols)})"`
which I don't love but seemed... good enough.

```r
mtcars_na <- mtcars
mtcars_na$mpg[1] <- NA
tidylog::arrange(mtcars_na, mpg, cyl) |> head(3)
#> arrange: sorted rows by mpg
#>          some columns contained NAs which sort last (mpg)
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4
#> Camaro Z28          13.3   8  350 245 3.73 3.840 15.41  0  0    3    4

**Comment by orgadish** [comment]
@elbersb I noticed that `test_join.R` incorrectly had `context("test_group_by")` at the top so group_by was showing up twice in the test results (and join not at all).

Are you ok with including the simple fix for that in this PR, or prefer for it to come in separately?

**Comment by elbersb** [comment]
Sure, ok to include here.

Re the message, how about something like this:

```r
tidylog::arrange(mtcars_na, mpg, cyl) |> head(3)
#> arrange: sorted rows by cyl, mpg (NAs at end), gear
```

**Comment by orgadish** [comment]
With one NA column, I think that makes sense. But with multiple NA columns, I think the line feels bloated and long. And then if there are more than 5 columns, I use `...` (using `format_list`), so a column with NA could be hidden.

Given that, I think it's probably better to keep those as separate lines. 

One approach could be to use an asterisk or other symbol in the first column and then use the second column to say "*: one or more column(s) contained NAs, which sort last" to avoid double-writing the columns. And if one of the hidden columns contains NA, the ellipsis can take an asterisk too.

**Comment by elbersb** [comment]
I like the asterisk approach, I'd probably phrase it something like `These columns contain NAs, which are sorted at the end`

**Comment by orgadish** [comment]
@elbersb I ended up using `cli::symbol$sup_1` because it makes it look like a footnote, which I think works well.

One thing I did notice is that since we don't currently process complicated column logic, we can't track if it does or doesn't have NAs (see `mpg * hp` or `mpg * 2` below). Should we add that logic? That feels like a slippery slope since that logic could be very slow and duplicated. We could also use `sup_2` to mark those differently and indicate that compound column definitions cannot be assessed for NAs?

``` r
library(dplyr)
devtools::load_all("D:/Github/og_tidylog")
#> ℹ Loading tidylog


mtcars_na <- mtcars
mtcars_na$mpg[1] <- NA
mtcars_na$gear[1] <- NA

tidylog::arrange(mtcars_na, mpg) |> head(2)
#> arrange: sorted rows by mpg¹
#> ¹contained NAs, which sort to the end
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, mpg, hp, gear) |> head(2)
#> arrange: sorted rows by mpg¹, hp, gear¹
#>          ¹contained NAs, which sort to the end
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, desc(mpg)) |> head(2)
#> arrange: sorted rows by desc(mpg)¹
#>          ¹contained NAs, which sort to the end
#>                 mpg cyl disp hp drat    wt  qsec vs am gear carb
#> Toyota Corolla 33.9   4 71.1 65 4.22 1.835 19.90  1  1    4    1
#> Fiat 128       32.4   4 78.7 66 4.08 2.200 19.47  1  1    4    1

tidylog::arrange(mtcars_na, across(contains("p"))) |> head(2)
#> arrange: sorted rows by mpg¹, disp, hp
#>          ¹contained NAs, which sort to the end
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4

tidylog::arrange(mtcars_na, across(contains("p"), desc)) |> head(2)
#> arrange: sorted rows by desc(mpg)¹, desc(disp), desc(hp)
#>          ¹contained NAs, which sort to the end
#>                 mpg cyl disp hp drat    wt  qsec vs am gear carb
#> Toyota Corolla 33.9   4 71.1 65 4.22 1.835 19.90  1  1    4    1
#> Fiat 128       32.4   4 78.7 66 4.08 2.200 19.47  1  1    4    1



# === Complex Logic Does Not Track NAs === #

tidylog::arrange(mtcars_na, mpg * 2) |> head(2)
#> arrange: sorted rows by mpg * 2
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, mpg * hp) |> head(2)
#> arrange: sorted rows by mpg * hp
#>              mpg cyl  disp hp drat    wt  qsec vs am gear carb
#> Merc 240D   24.4   4 146.7 62 3.69 3.190 20.00  1  0    4    2
#> Honda Civic 30.4   4  75.7 52 4.93 1.615 18.52  1  1    4    2
```

<sup>Created on 2026-06-20 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by orgadish** [comment]
@elbersb What do you think of this in the latest update? I use `sup_1` for NAs ("contains NAs, which sort to the end" and `sup_2` for "NA status unknown" in the case where we can't know without evaluating. Note that the `sup_1` and `sup_2` assignment is fixed, even if the columns show up in the reverse order, so it doesn't show up the way footnotes would. But fixing this would require a decent amount of work to track the markers and then check the order at the end, which I don't think is worth it for these corner cases.

(The latest changes also fixed the `.data$col` case which is now displayed as `col`.)

I find the best way to check that the behavior is as desired for all the random cases is to check all the tests.

Let me know what you think! I think this is ready to merge if you agree with the design and wording.

``` r
suppressMessages(suppressWarnings(library(dplyr)))
devtools::load_all("D:/Github/og_tidylog")
#> ℹ Loading tidylog


mtcars_na <- mtcars
mtcars_na$mpg[1] <- NA
mtcars_na$gear[1] <- NA


# === Complex Logic Cannot Track NAs === #
tidylog::arrange(mtcars_na, mpg, hp, mpg * 2) |> head(2)
#> arrange: sorted rows by mpg¹, hp, mpg * 2²
#>          ¹contains NAs, which sort to the end
#>          ²NA status unknown
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, mpg * 2, hp, mpg) |> head(2)
#> arrange: sorted rows by mpg * 2², hp, mpg¹
#>          ¹contains NAs, which sort to the end
#>          ²NA status unknown
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, mpg * hp) |> head(2)
#> arrange: sorted rows by mpg * hp²
#>          ²NA status unknown
#>              mpg cyl  disp hp drat    wt  qsec vs am gear carb
#> Merc 240D   24.4   4 146.7 62 3.69 3.190 20.00  1  0    4    2
#> Honda Civic 30.4   4  75.7 52 4.93 1.615 18.52  1  1    4    2
```

<sup>Created on 2026-06-20 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>


---
```r
> devtools::test()
ℹ Testing tidylog
✔ | F W  S  OK | Context
✔ |         90 | test_arrange                                                                                                             
✔ |         85 | test_filter                                                                                                              
✔ |         12 | test_group_by                                                                                                            
✔ |         84 | test_join [4.1s]                                                                                                         
✔ |         56 | test_longer_wider                                                                                                        
✔ |         75 | test_mutate                                                                                                              
✔ |         13 | test_rename                                                                                                              
✔ |         25 | test_select                                                                                                              
✔ |         30 | test_summarize                                                                                                           
✔ |         16 | test_tidylog                                                                                                             

══ Results ═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 6.8 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 486 ]

**Comment by elbersb** [comment]
I'll let you make the call, but this is getting awfully complicated I think? The limitations around NA handling make me think that maybe we want to just not have this for now (we could reconsider if a requests comes in?). Having up to three log lines for a fairly simple operation just seems like a bit too much?

**Comment by orgadish** [comment]
I agree the NA handling makes it much more complicated. And since it relies on syntax parsing, it's more directly coupled to dplyr behavior and rules than the other wrappers.

However, this was already a suggested request and I think the current implementation is how I think we would do it anyway. So I think we should keep it. Any new behavior from dplyr would get caught by the fallback and just not have proper NA handling, which is fine.

Since the lines are kinda like footnotes anyway, I can combine the sup_1; sup_2 lines into a single, semi-colon separated line. That will reduce the noise.

**Comment by orgadish** [comment]
e90f058 and b323864 moved it to be on one line and wrapped complex variables in backticks. I think that makes this cleaner, less noisy, and nicer looking. I think this is a good place to leave it!

``` r
suppressMessages(suppressWarnings(library(dplyr)))
devtools::load_all("D:/Github/og_tidylog")
#> ℹ Loading tidylog


mtcars_na <- mtcars
mtcars_na$mpg[1] <- NA
mtcars_na$gear[1] <- NA


# === Complex Logic Cannot Track NAs === #
tidylog::arrange(mtcars_na, mpg, hp, is.na(mpg)) |> head(2)
#> arrange: sorted rows by mpg¹, hp, `is.na(mpg)`²
#>          ¹contains NAs, which sort to the end; ²NA status unknown
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4

tidylog::arrange(mtcars_na, is.na(mpg), hp, mpg) |> head(2)
#> arrange: sorted rows by `is.na(mpg)`², hp, mpg¹
#>          ¹contains NAs, which sort to the end; ²NA status unknown
#>              mpg cyl  disp hp drat    wt  qsec vs am gear carb
#> Honda Civic 30.4   4  75.7 52 4.93 1.615 18.52  1  1    4    2
#> Merc 240D   24.4   4 146.7 62 3.69 3.190 20.00  1  0    4    2

tidylog::arrange(mtcars_na, mpg * hp) |> head(2)
#> arrange: sorted rows by `mpg * hp`²
#>          ²NA status unknown
#>              mpg cyl  disp hp drat    wt  qsec vs am gear carb
#> Merc 240D   24.4   4 146.7 62 3.69 3.190 20.00  1  0    4    2
#> Honda Civic 30.4   4  75.7 52 4.93 1.615 18.52  1  1    4    2
```

<sup>Created on 2026-06-23 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

File: DESCRIPTION [hunk 1]
```diff
@@ -15,7 +15,8 @@ Imports:
     tidyr,
     glue,
     cli,
-    rlang (>= 0.4.3)
+    rlang (>= 0.4.3),
+    tidyselect
```
Comment (orgadish) [inline DESCRIPTION:19]: @elbersb Note that this does not actually add a new dependency, since `tidyselect` is required by `dplyr` already anyway

**Comment by orgadish** [comment]
@elbersb I merged to resolve conflicts from master, and I think this is now ready to merge, if you agree with the design.

**Comment by orgadish** [comment]
Updated tests to remove new helpers after discussion on r-lib/testthat#2342 showed that you could pipe `expect_message()` together as long as you use the `message = ...` named argument.

**Comment by elbersb** [comment]
Hi, sorry, just getting back to this. Should I merge?

**Comment by orgadish** [comment]
@elbersb I just took a quick look again at the code and it looks right (I hadn't touched or looked at it since my last comment). 

I think it's good to merge!

File: R/arrange.R [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+# Note: on non-UTF8 terminals, cli falls back to plain digits ("1"/"2")
+# for these symbols.
+log_arrange <- function(.olddata, .newdata, .funname, ...) {
+    if (!"data.frame" %in% class(.olddata) || !should_display()) {
+        return()
+    }
+
+    is_grouped <- dplyr::is_grouped_df(.olddata)
+    grp_prefix <- if (is_grouped) " (grouped)" else ""
+    grp_infix <- if (is_grouped) " within groups" else ""
```
Comment (elbersb) [inline R/arrange.R:40]: Merge blocker: grouped data is always described as sorted “within groups,” but `dplyr::arrange()` sorts grouped data globally unless `.by_group = TRUE`. For example, sorting groups `a,b,a,b` by values `2,1,1,2` returns group order `b,a,a,b` while this message says “within groups.” Make this text depend on `.by_group`.

File: R/arrange.R [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    } else {
+        args[[1]]
+    }
+
+    resolved <- resolve_tidyselect(inner, data, env)
+
+    has_desc_fn <- length(args) >= 2 &&
+        rlang::expr_text(args[[2]]) %in% c("desc", "dplyr::desc")
+
+    labels <- if (has_desc_fn) paste0("desc(", resolved, ")") else resolved
```
Comment (elbersb) [inline R/arrange.R:167]: Merge blocker: valid `across()` transformations other than exact `desc` are reported as bare-column sorts. `arrange(data.frame(x = c(1,3,2)), across(x, ~ -.x))` returns `3,2,1` but logs `sorted rows by x`. Report the transformation, or mark the expression as unknown.

File: R/arrange.R [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+# Evaluates a column selection expression against data, returning matched column names.
+# Handles tidyselect helpers (everything(), starts_with(), all_of(), c(), a:b, etc.)
+# Does not automatically unpack across() and pick(), for example.
+# Returns character(0) if the expression is not a valid column selection.
+resolve_tidyselect <- function(expr, data, env) {
+    # Must pass the quosure environment down to ensure that the correct
+    # variable scope is available, e.g. for `all_of(vars)`.
+    tryCatch(
+        names(tidyselect::eval_select(expr, data, env = env)),
```
Comment (elbersb) [inline R/arrange.R:131]: Merge blocker: this evaluates the user selector a second time after `dplyr::arrange()` already evaluated it. Verified with `n <- 0; arrange(df, across(all_of({n <<- n + 1; "x"})))`: `n` becomes 2. Logging must not repeat user side effects.

File: R/arrange.R [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        return()
+    }
+
+    # Capture as quosures to preserve the user's environment
+    vars <- rlang::enquos(...)
+
+    # Named arguments to arrange() that are options, not sorting columns.
+    # Update if dplyr adds new named arguments to arrange().
+    to_exclude <- c(".by_group", ".locale")
+    vars <- vars[!names(vars) %in% to_exclude]
```
Comment (elbersb) [inline R/arrange.R:56]: Merge blocker: `.by_group = TRUE` can change row order without any sort expression, but removing it leaves no labels. A grouped input ordered `b,a` is reordered to `a,b` and logs `sorted rows within groups by ` with an empty suffix. Handle grouping columns when `.by_group` causes the sort.

File: R/arrange.R (hunk 3 above)
Comment (elbersb) [reply R/arrange.R:167]: not sure we can support all use cases here, but this one should probably be marked as unknown?

File: R/arrange.R (hunk 2 above)
Comment (elbersb) [reply R/arrange.R:40]: this sounds valid to me

File: R/arrange.R (hunk 4 above)
Comment (elbersb) [reply R/arrange.R:131]: not sure this will ever be a problem in practice, but I guess technically it's true

**Comment by elbersb** [comment]
Because this is quite a lot of code I had an LLM look over it again, and it found 4 things. Not sure all of these are extremely important to fix, but especially the first one sounds like a bug?

File: R/arrange.R (hunk 5 above)
Comment (elbersb) [reply R/arrange.R:56]: bit of an edge maybe, but valid
